pep-0xxx.rst

```diff
@@ -0,0 +1,727 @@
+            during the TLS handshake. This may be advertised using either or
+            both of ALPN or NPN.
+
+            ``protocols`` should be a list of acceptable protocols in the form
+            of ``NextProtocol`` objects, such as ``[H2, HTTP1]``, ordered by
+            preference. The selection of the protocol will happen during the
+            handshake, and will use whatever protocol negotiation mechanisms
+            are available and supported by both peers.
+
+            If the TLS implementation doesn't support protocol negotiation,
+            this method will raise ``NotImplementedError``.
+            """
+
+        @abstractmethod
+        def set_sni_callback(self, callback: Optional[ServerNameCallback]) -> None:
+            """
+            Register a callback function that will be called after the TLS
+            Client Hello handshake message has been received by the TLS server
+            when the TLS client specifies a server name indication.
+
+            Only one callback can be set per ``Context``. If ``callback`` is
+            ``None`` then the callback is disabled. Calling this function a
+            subsequent time will disable the previously registered callback.
+
+            The ``callback`` function will be called with three arguments: the
+            first will be the ``TLSBufferObject`` for the connection; the
+            second will be a string that represents the server name that the
+            client is intending to communicate (or ``None`` if the TLS Client
+            Hello does not contain a server name); and the third argument will
+            be the original ``Context``. The server name argument will be the
+            IDNA *decoded* server name.
+
+            The ``callback`` must return ``None`` to allow negotiation to
+            continue. Other return values signal errors. Attempting to control
+            what error is signaled by the underlying TLS implementation is not
+            specified in this API, but is up to the concrete implementation to
+            handle.
+            """
+
+        @abstractmethod
+        def set_version_range(self, lower_bound=None: Optional[TLSVersion],
+                              upper_bound=None: Optional[TLSVersion]) -> None:
+            """
+            Set the minumum and maximum versions of TLS that should be allowed
+            on TLS connections made by this context.
+
+            If present, ``lower_bound`` will set the lowest acceptable TLS
+            version. If present, ``upper_bound`` will set the highest
+            acceptable TLS version. If either argument is ``None``, this will
+            leave that bound unchanged.
+            """
+
+        @abstractmethod
+        def wrap_socket(self, socket: socket.socket, server_side=False: bool,
+                        auto_handshake=True: bool,
+                        server_hostname=None: Optional[str]) -> TLSWrappedSocket:
```

Comment:
> So... either an option or only-False seem reasonable to me, but either way I think the spec needs to explicitly say what it's doing :-).

I remain unconvinced of the need to do this up here. I can see wanting to provide an implementers note, but ultimately this seems like a really weird note to give. It's essentially saying: "be aware that the connection may be terminated in an unclean manner, and that may be treated as an error condition". This document really shouldn't become a "this is how TLS works" primer.

We absolutely *should* consider adding a section ("Porting Notes"?) that covers gotchas that may be encountered when moving from the stdlib `ssl` module to this one. But I don't think annotating each method that differs from stdlib `ssl` in any minor way about exactly how it differs is a sensible way of presenting the documentation of the interface.

For `session`, the big risk here is that in an *abstract API* persisting session information is a very tricky thing to do. This is because it is clearly not enough to persist an abstract session ID, as that would provide little value across different processes. You need to persist all relevant information about the Session, and what exactly that data is and how it gets turned into a Session that a concrete implementation can restore is *extremely* esoteric. Essentially, if you're using the *generic* API, the idea of doing cross-process session caching is something you need to give up on, because you simply cannot meaningfully assume that your OpenSSL-based session can be quietly inserted into a GnuTLS-based process.

This is a problem for concrete APIs, and they can address it directly in their own implementations.